File: docs/source/src/python/user-guide/concepts/data-types-and-structures.py
# --8<-- [start:series]
import polars as pl

s = pl.Series("ints", [1, 2, 3, 4, 5])
print(s)
# --8<-- [end:series]

# --8<-- [start:series-dtype]
s1 = pl.Series("ints", [1, 2, 3, 4, 5])
s2 = pl.Series("uints", [1, 2, 3, 4, 5], dtype=pl.UInt64)
print(s1.dtype, s2.dtype)
# --8<-- [end:series-dtype]

# --8<-- [start:df]
from datetime import date

df = pl.DataFrame(
    {
        "name": ["Alice Archer", "Ben Brown", "Chloe Cooper", "Daniel Donovan"],
        "birthdate": [
            date(1997, 1, 10),
            date(1985, 2, 15),
            date(1983, 3, 22),
            date(1981, 4, 30),
        ],
        "weight": [57.9, 72.5, 53.6, 83.1],  # (kg)
        "height": [1.56, 1.77, 1.65, 1.75],  # (m)
    }
)

print(df)
# --8<-- [end:df]

# --8<-- [start:schema]
print(df.schema)
# --8<-- [end:schema]

# --8<-- [start:head]
print(df.head(3))
# --8<-- [end:head]

# --8<-- [start:glimpse]
print(df.glimpse(return_as_string=True))
# --8<-- [end:glimpse]

# --8<-- [start:tail]
print(df.tail(3))
# --8<-- [end:tail]

# --8<-- [start:sample]
import random

random.seed(42)  # For reproducibility.

print(df.sample(2))
# --8<-- [end:sample]

# --8<-- [start:describe]
print(df.describe())
# --8<-- [end:describe]

File: docs/source/src/rust/user-guide/concepts/data-types-and-structures.rs
fn main() {
    // --8<-- [start:series]
    use polars::prelude::*;

    let s = Series::new("ints".into(), &[1, 2, 3, 4, 5]);

    println!("{}", s);
    // --8<-- [end:series]

    // --8<-- [start:series-dtype]
    let s1 = Series::new("ints".into(), &[1, 2, 3, 4, 5]);
    let s2 = Series::new("uints".into(), &[1, 2, 3, 4, 5])
        .cast(&DataType::UInt64) // Here, we actually cast after inference.
        .unwrap();
    println!("{} {}", s1.dtype(), s2.dtype()); // i32 u64
                                               // --8<-- [end:series-dtype]

    // --8<-- [start:df]
    use chrono::prelude::*;

    let df: DataFrame = df!(
        "name" => ["Alice Archer", "Ben Brown", "Chloe Cooper", "Daniel Donovan"],
        "birthdate" => [
            NaiveDate::from_ymd_opt(1997, 1, 10).unwrap(),
            NaiveDate::from_ymd_opt(1985, 2, 15).unwrap(),
            NaiveDate::from_ymd_opt(1983, 3, 22).unwrap(),
            NaiveDate::from_ymd_opt(1981, 4, 30).unwrap(),
        ],
        "weight" => [57.9, 72.5, 53.6, 83.1],  // (kg)
        "height" => [1.56, 1.77, 1.65, 1.75],  // (m)
    )
    .unwrap();
    println!("{}", df);
    // --8<-- [end:df]

    // --8<-- [start:schema]
    println!("{:?}", df.schema());
    // --8<-- [end:schema]

    // --8<-- [start:head]
    let df_head = df.head(Some(3));

    println!("{}", df_head);
    // --8<-- [end:head]

    // --8<-- [start:tail]
    let df_tail = df.tail(Some(3));

    println!("{}", df_tail);
    // --8<-- [end:tail]

    // --8<-- [start:sample]
    let n = Series::new("".into(), &[2]);
    let sampled_df = df.sample_n(&n, false, false, None).unwrap();

    println!("{}", sampled_df);
    // --8<-- [end:sample]

    // --8<-- [start:describe]
    // Not available in Rust
    // --8<-- [end:describe]
}
